### PR TITLE
feat: Add grafana Promtail to Kubernetes addon

### PR DESCRIPTION
**Pull request by florentio**
### What does this PR do?
Issue https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/774

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [x] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [x] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: docs/add-ons/promtail.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# PROMTAIL
```
Comment (bryantbiggs) [inline docs/add-ons/promtail.md:1]:
```suggestion
# Promtail
```

File: docs/add-ons/promtail.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+Promtail is an agent which ships the contents of local logs to a Loki instance.
+
+[PROMTAIL](https://github.com/grafana/helm-charts/tree/main/charts/promtail) chart bootstraps Promtail infrastructure on a Kubernetes cluster using the Helm package manager.
```
Comment (bryantbiggs) [inline docs/add-ons/promtail.md:5]:
```suggestion
[Promtail](https://github.com/grafana/helm-charts/tree/main/charts/promtail) chart bootstraps Promtail infrastructure on a Kubernetes cluster using the Helm package manager.
```

File: docs/add-ons/promtail.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+For complete project documentation, please visit the [Promtail documentation site](https://grafana.com/docs/loki/latest/clients/promtail/).
+
+## Usage
+
+PROMTAIL can be deployed by enabling the add-on via the following.
```
Comment (bryantbiggs) [inline docs/add-ons/promtail.md:11]:
```suggestion
PROMTAIL can be deployed by enabling the add-on via the following.
```

Not sure why these are all caps but lets just use title case on all of these `Promtail`

File: docs/add-ons/promtail.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+```hcl
+enable_promtail = true
+```
+
+Deploy PROMTAIL with custom `values.yaml`
+
+```hcl
+  # Optional Map value; pass promtail-values.yaml from consumer module
+  promtail_helm_config = {
+    name       = "promtail"                                               # (Required) Release name.
```
Comment (bryantbiggs) [inline docs/add-ons/promtail.md:22]: Release name is not required, but chart name is.

File: docs/add-ons/promtail.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    namespace  = "promtail"                                               # (Optional) The namespace to install the release into.
+    values = [templatefile("${path.module}/promtail-values.yaml", {})]
+  }
+```
+
+### GitOps Configuration
+
+The following properties are made available for use when managing the add-on via GitOps.
+
+```
```
Comment (bryantbiggs) [inline docs/add-ons/promtail.md:35]:
```suggestion
```hcl
```

File: modules/kubernetes-addons/promtail/locals.tf [hunk 6]
```diff
@@ -0,0 +1,23 @@
+locals {
+  name = "promtail"
+  default_helm_config = {
+    name             = local.name
+    chart            = local.name
+    repository       = "https://grafana.github.io/helm-charts"
+    version          = "6.3.0"
+    namespace        = try(var.helm_config["namespace"], local.name)
+    timeout          = "1200"
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/promtail/locals.tf:9]: Lets remove, the Helm addon has this as a default already

File: modules/kubernetes-addons/promtail/locals.tf [hunk 7]
```diff
@@ -0,0 +1,23 @@
+locals {
+  name = "promtail"
+  default_helm_config = {
+    name             = local.name
+    chart            = local.name
+    repository       = "https://grafana.github.io/helm-charts"
+    version          = "6.3.0"
+    namespace        = try(var.helm_config["namespace"], local.name)
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/promtail/locals.tf:8]: Since this is getting merged anyways, this is not needed like this
```suggestion
    namespace        = local.name
```

File: modules/kubernetes-addons/promtail/variables.tf [hunk 8]
```diff
@@ -0,0 +1,28 @@
+variable "helm_config" {
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/promtail/variables.tf:1]: These should all be ordered
```hcl
variable "foo" {
  description = "First"
  type        = any
  default     = {}
}
```

**Review by bryantbiggs — COMMENTED** [review]
Overall looks great - few minor comments.

Lets add an example for this - we'll call it `grafana-loki` - this is just the start where we have promtail and later we can add loki and connect the two. 

You can copy this [example](https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/main/examples/stateful/main.tf) and remove the EFS and Velero bits

File: examples/grafana-loki/versions.tf [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+      source  = "hashicorp/random"
+      version = ">= 3.0"
+    }
+  }
+
+  # ##  Used for end-to-end testing on project; update to suit your needs
+  # backend "s3" {
+  #   bucket = "terraform-ssp-github-actions-state"
+  #   region = "us-west-2"
+  #   key    = "e2e/stateful/terraform.tfstate"
```
Comment (bryantbiggs) [inline examples/grafana-loki/versions.tf:27]:
```suggestion
  #   key    = "e2e/grafana-loki/terraform.tfstate"
```

**Review by bryantbiggs — APPROVED** [review]
This looks great, thank you!

**Comment by sondrelg** [comment]
I've just implemented this (thanks for adding it, it's super helpful; excited to see loki being added in the future), but I'm running into an issue where the `promtail_helm_config` values are not being picked up by terraform.

Is it possible that the loading of `values` from the `promtail_helm_config` object is broken?

This is my configuration (which I've tweaked several ways to try and get it to work, without luck):

```yaml
  enable_promtail = true
  promtail_helm_config = {
    name       = "promtail"
    repository = "https://grafana.github.io/helm-charts"
    chart      = "promtail"
    version    = "6.3.0"
    namespace  = "promtail"
    values = [templatefile("${path.module}/promtail-values.yml", {})]
  }
```

My `promtail-values.yml` contains:
```yaml
config:
  clients:
    - url: https://url-to-another-cluster.com/loki/api/v1/push
```

------------------------

tl;dr: despite setting a URL value, the pods try to connect to the default value (`http://loki-gateway/loki/api/v1/push`) instead, and when I make changes to my config, the changes are not detected by terraform.

When I replace the above config with something nonsensical like `aasdasd`, the apply still runs fine. I'm pretty sure other addons would fail with a yaml-parsing error. This increases my suspicion that the contents of the file just aren't evaluated.

Is there a chance that something needs tweaking wrt. how the helm config is evaluated for this addon?

**Comment by sondrelg** [comment]
Let me know if you prefer I open an issue 🙏 Just thought it might be nice to mention it here first, since both of you might be able to answer relatively quickly.

**Comment by sondrelg** [comment]
For some added context, the config values are applied correctly when I use a normal helm release:
```yaml
resource "helm_release" "loki" {
    name       = "promtail"
    repository = "https://grafana.github.io/helm-charts"
    chart      = "promtail"
    version    = "6.3.0"
    namespace  = "promtail"
    values = [templatefile("${path.module}/promtail-values.yml", {})]
}
```
Which also seems to point in the direction that something is wrong with addon loading (or my definition of it perhaps?)

**Comment by florentio** [comment]
Hello @sondrelg I couldn't reproduce your error, this is what I got where running terraform plan with your config

```hcl
# module.eks.module.promtail.module.helm_addon.helm_release.addon[0] will be created
  + resource "helm_release" "addon" {
      + atomic                     = false
      + chart                      = "promtail"
      + cleanup_on_fail            = false
      + create_namespace           = true
      + dependency_update          = false
      + description                = "Promtail helm Chart deployment configuration"
      + disable_crd_hooks          = false
      + disable_openapi_validation = false
      + disable_webhooks           = false
      + force_update               = false
      + id                         = (known after apply)
      + lint                       = false
      + manifest                   = (known after apply)
      + max_history                = 0
      + metadata                   = (known after apply)
      + name                       = "promtail"
      + namespace                  = "promtail"
      + pass_credentials           = false
      + recreate_pods              = false
      + render_subchart_notes      = true
      + replace                    = false
      + repository                 = "https://grafana.github.io/helm-charts"
      + reset_values               = false
      + reuse_values               = false
      + skip_crds                  = false
      + status                     = "deployed"
      + timeout                    = 1200
      + values                     = [
          + <<-EOT
                config:
                  clients:
                    - url: https://url-to-another-cluster.com/loki/api/v1/push
            EOT,
        ]
      + verify                     = false
      + version                    = "6.3.0"
      + wait                       = true
      + wait_for_jobs              = false

      + postrender {}
    }

````

and as you can see the pod is using `https://url-to-another-cluster.com/loki/api/v1/push` as url  in the log file below
[promtail-xq9xv.log](https://github.com/aws-ia/terraform-aws-eks-blueprints/files/9496441/promtail-xq9xv.log)

Can you open an issue and share the terraform plan section of `resource "helm_release" "addon" ` so that we can see the plan content ?

**Comment by sondrelg** [comment]
Thanks for testing @florentio. I'll double check and report back in a bit 👍
